eras/babbage/impl/src/Cardano/Ledger/Babbage.hs

```diff
@@ -0,0 +1,281 @@
+import Cardano.Ledger.SafeHash (hashAnnotated)
+import Cardano.Ledger.Shelley (nativeMultiSigTag)
+import qualified Cardano.Ledger.Shelley.API as API
+import Cardano.Ledger.Shelley.Constraints
+  ( UsesPParams (..),
+    UsesTxOut (..),
+    UsesValue,
+  )
+import Cardano.Ledger.Shelley.EpochBoundary
+import Cardano.Ledger.Shelley.Genesis (genesisUTxO, sgGenDelegs, sgMaxLovelaceSupply, sgProtocolParams)
+import Cardano.Ledger.Shelley.LedgerState
+  ( AccountState (..),
+    DPState (..),
+    EpochState (..),
+    LedgerState (..),
+    NewEpochState (..),
+    smartUTxOState,
+    _genDelegs,
+  )
+import Cardano.Ledger.Shelley.Metadata (validMetadatum)
+import qualified Cardano.Ledger.Shelley.Rules.Epoch as Shelley
+import qualified Cardano.Ledger.Shelley.Rules.Mir as Shelley
+import qualified Cardano.Ledger.Shelley.Rules.Newpp as Shelley
+import qualified Cardano.Ledger.Shelley.Rules.Rupd as Shelley
+import qualified Cardano.Ledger.Shelley.Rules.Snap as Shelley
+import qualified Cardano.Ledger.Shelley.Rules.Tick as Shelley
+import qualified Cardano.Ledger.Shelley.Rules.Upec as Shelley
+import qualified Cardano.Ledger.Shelley.Tx as Shelley
+import Cardano.Ledger.Shelley.UTxO (balance)
+import Cardano.Ledger.ShelleyMA.Rules.Utxo (consumed)
+import Cardano.Ledger.ShelleyMA.Timelocks (validateTimelock)
+import Cardano.Ledger.Val (Val (inject), coin, (<->))
+import Control.Arrow (left)
+import Control.Monad.Except (liftEither)
+import Control.Monad.Reader (runReader)
+import Control.State.Transition.Extended (TRC (TRC))
+import Data.Default (def)
+import qualified Data.Map.Strict as Map
+import Data.Maybe.Strict
+import qualified Data.Set as Set
+
+-- =====================================================
+
+-- | The Alonzo era
+data BabbageEra c
+
+instance
+  ( CC.Crypto c,
+    era ~ BabbageEra c
+  ) =>
+  EraModule.Era (BabbageEra c)
+  where
+  type Crypto (BabbageEra c) = c
+
+  getTxOutEitherAddr = getBabbageTxOutEitherAddr
+
+instance API.ShelleyEraCrypto c => API.ApplyTx (BabbageEra c) where
+  reapplyTx globals env state vtx =
+    let res =
+          flip runReader globals
+            . applySTSNonStatic
+              @(Core.EraRule "LEDGER" (BabbageEra c))
+            $ TRC (env, state, API.extractTx vtx)
+     in liftEither . left API.ApplyTxError $ res
+
+instance API.ShelleyEraCrypto c => API.ApplyBlock (BabbageEra c)
+
+instance (CC.Crypto c) => Shelley.ValidateScript (BabbageEra c) where
+  isNativeScript x = not (isPlutusScript x)
+  scriptPrefixTag script =
+    case script of
+      (TimelockScript _) -> nativeMultiSigTag -- "\x00"
+      (PlutusScript PlutusV1 _) -> "\x01"
+      (PlutusScript PlutusV2 _) -> "\x02"
+  validateScript (TimelockScript script) tx = validateTimelock @(BabbageEra c) script tx
+  validateScript (PlutusScript _ _) _tx = True
+
+instance
+  ( CC.Crypto c
+  ) =>
+  API.CanStartFromGenesis (BabbageEra c)
+  where
+  type AdditionalGenesisConfig (BabbageEra c) = AlonzoGenesis
```

Comment:
I need to remind myself of the details about how this works to convince myself that `AlonzoGenesis` makes sense here...

Reply:
it seems like `AlonzoGenesis` exists so we can encode/decode the arguments to `extendPParams`, nothing alonzo-specific gets encoded so 🤷